### ISSUE TITLE
GraphQL - Cannot read property 'isMongoId' of undefined

### ISSUE THREAD

**Issue by gianniskarmas** [strapi/strapi#2408]
**Informations**
- **Node.js version**: v10.13.0
- **NPM version**: 6.4.1
- **Strapi version**: 3.0.0-alpha.15
- **Database**: Mongo
- **Operating system**: OSX
**What is the current behavior?**
in a Many to Many relation GraphQl on latest strapi throws:
> "Cannot read property 'isMongoId' of undefined"
> "at valueToId (/Users/admin/Downloads/kon-strapi/node_modules/strapi-hook-mongoose/lib/utils/index.js:60:19)"
> 

**Steps to reproduce the problem**
Create two Collections and a many to many field, add a couple a values and query with graphql:
```graphql
{
  categories {
    name
    Product{
      Name
    }
  }
}
```

**What is the expected behavior?**
To return the results.
In 14.5 it was working fine but now it is not.

**Suggested solutions**
Couldn't find any. Seems relevant to issue [#1683](https://github.com/strapi/strapi/issues/1683#issue-346752826)

**Comment by gianniskarmas** [comment]
It also has something to do with 'dominant': true.
If you query the other way around it works sometimes.

**Comment by mamartins** [comment]
> It also has something to do with 'dominant': true.
> If you query the other way around it works sometimes.

Precisely! The dominant side has the problem here

**Comment by mamartins** [comment]
This rework under the `strapi-hook-mongoose/lib/utils` solves the error, but the relation returns always empty:
```
'use strict';

/**
 * Module dependencies
 */

// Public node modules.
const mongoose = require('mongoose');
const Mongoose = mongoose.Mongoose;

/**
 * Convert MongoDB ID to the stringify version as GraphQL throws an error if not.
 *
 * Refer to: https://github.com/graphql/graphql-js/commit/3521e1429eec7eabeee4da65c93306b51308727b#diff-87c5e74dd1f7d923143e0eee611f598eR183
 */
mongoose.Types.ObjectId.prototype.valueOf = function () {
  return this.toString();
};

module.exports = (mongoose = new Mongoose()) => {

  const Decimal = require('mongoose-float').loadType(mongoose, 2);
  const Float = require('mongoose-float').loadType(mongoose, 20);

  const r = {
    convertType: mongooseType => {
      switch (mongooseType.toLowerCase()) {
        case 'array':
          return Array;
        case 'boolean':
          return 'Boolean';
        case 'binary':
          return 'Buffer';
        case 'date':
        case 'datetime':
        case 'time':
        case 'timestamp':
          return Date;
        case 'decimal':
          return Decimal;
        case 'float':
          return Float;
        case 'json':
          return 'Mixed';
        case 'biginteger':
        case 'integer':
          return 'Number';
        case 'uuid':
          return 'ObjectId';
        case 'email':
        case 'enumeration':
        case 'password':
        case 'string':
        case 'text':
          return 'String';
        default:
      }
    },
    valueToId: function (value) {
      return r.isMongoId(value)
        ? mongoose.Types.ObjectId(value)
        : value;
    },
    isMongoId: function (value) {
      // Here we don't use mongoose.Types.ObjectId.isValid method because it's a weird check,
      // it returns for instance true for any integer value ¯\_(ツ)_/¯
      const hexadecimal = /^[0-9A-F]+$/i;
      return hexadecimal.test(value) && `${value}`.length === 24;
    }
  };

  return r;
};
```

**Comment by mamartins** [comment]
@lauriejim Don't you need the validation that I've fixed at `strapi-hook-mongoose/lib/utils` on my previous post?

**Comment by gianniskarmas** [comment]
BTW, this `const hexadecimal = /^[0-9A-F]+$/i;` looks wrong.
It should contain lowercase too `/^[0-9a-fA-F]+$/i` right?

**Comment by lauriejim** [comment]
I close this issue. It was related to a PR released on `alpha.15` and we reverted it and pusblish new version of Strapi.

**Comment by mamartins** [comment]
fixed on alpha 16?

**Comment by ByBogon** [comment]
I also have kinda similar issue on alpha 24.1. is that fixed now on alpha 24.1?
